### PR TITLE
CM_APP_55 Recover CM Id

### PR DESCRIPTION
**Pull request by gayatrijagtap**

### CODE REVIEW

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/UserService.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/ABPMJAYIdFilter.java  4
- src/main/java/in/projecteka/consentmanager/user/NameFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/YOBFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/PatientMatcher.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/filters/YOBFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/IdentifierUtils.java  1
- src/main/java/in/projecteka/consentmanager/user/filters/NameFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/filters/ABPMJAYIdFilter.java  4
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/filters/YOBFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/filters/ABPMJAYIdFilter.java  4
- src/main/java/in/projecteka/consentmanager/user/IdentifierUtils.java  1
- src/main/java/in/projecteka/consentmanager/user/filters/NameFilter.java  2
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

File: src/main/java/in/projecteka/consentmanager/clients/ClientError.java [hunk 1]
```diff
@@ -308,4 +310,20 @@ public static ClientError unprocessableEntity() {
...
+        return new ClientError(BAD_REQUEST,
+                new ErrorRepresentation(new Error(INVALID_RECOVERY_REQUEST, "Invalid CM Id recovery request")));
+    }
+
+    public static ClientError noPatientFound() {
+        return new ClientError(NOT_FOUND,
+                new ErrorRepresentation(new Error(NO_PATIENT_FOUND, "No patient matching the records")));
+    }
+
+    public static ClientError multiplePatientsFound() {
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/clients/ClientError.java:323]: I think we should not reveal any additional info regarding whether multiple patients were identified in the process of recovery. Taking other system references for recovery, none says so in the recovery process. I think we should throw a generic message saying - can not identify patient, insufficient info and also not send out a specific MULTIPLE_PATIENTS_FOUND error. Check with Akshatha.

File: src/main/java/in/projecteka/consentmanager/user/UserRepository.java [hunk 2]
```diff
@@ -21,6 +25,9 @@
     private static final String SELECT_PATIENT = "select id, name, gender, year_of_birth, phone_number, unverified_identifiers " +
             "from patient where id = $1";
 
+    private static final String SELECT_PATIENT_BY_DETAILS = "select id, year_of_birth, unverified_identifiers, name from patient" +
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/user/UserRepository.java:28]: rename - SELECT_PATIENT_BY_GENDER_MOB ?

File: src/main/java/in/projecteka/consentmanager/user/UserService.java [hunk 3]
```diff
@@ -229,4 +237,36 @@ public String getUserIdSuffix() {
...
+    }
+
+    private Mono<RecoverCmIdRequest> isInvalidRecoveryRequest(RecoverCmIdRequest request) { //breakdown
+        boolean areMandatoryFieldsNull = request.getName() == null || request.getGender() == null || !IdentifierUtils.isIdentifierTypePresent(request.getVerifiedIdentifiers(), IdentifierType.MOBILE);
+        boolean isInvalidVerifiedIdentifierMapped = isInvalidIdentifierMapped(request.getVerifiedIdentifiers(), IdentifierGroup.VERIFIED_IDENTIFIER);
+        boolean isInvalidUnverifiedIdentifierMapped = isInvalidIdentifierMapped(request.getUnverifiedIdentifiers(), IdentifierGroup.UNVERIFIED_IDENTIFIER);
+        return areMandatoryFieldsNull || isInvalidVerifiedIdentifierMapped || isInvalidUnverifiedIdentifierMapped ? Mono.empty() : Mono.just(request);
+    }
+
+    private Mono<User> validateNonZeroRows(List<User> rows) {
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/user/UserService.java:269]: if you are returning a Mono<User>, maybe consider renaming the method - getDistinctUser( .. )

File: src/main/java/in/projecteka/consentmanager/user/model/IdentifierType.java [hunk 4]
```diff
@@ -16,4 +18,15 @@ public boolean isValid(String abpmJayValue) {
     };
 
     public abstract boolean isValid(String value);
+
+    public static final HashMap<IdentifierType, IdentifierGroup> identifierTypeGroupMap = new HashMap<>() {
```
Comment (angshu) [inline src/main/java/in/projecteka/consentmanager/user/model/IdentifierType.java:22]: I would have preferred to keep this mapping external. You already have a Utils class, do you want to move it there? In the future, this mapping should be externalized.

**Comment by angshu** [comment]
Also check the sonar cloud analysis.

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/IdentifierUtils.java  1
- src/main/java/in/projecteka/consentmanager/user/filters/NameFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/filters/YOBFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/filters/ABPMJAYIdFilter.java  4
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

File: src/main/java/in/projecteka/consentmanager/user/UserRepository.java (hunk 2 above)
Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/UserRepository.java]: done

File: src/main/java/in/projecteka/consentmanager/user/UserService.java (hunk 3 above)
Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/UserService.java]: renamed

File: src/main/java/in/projecteka/consentmanager/user/model/IdentifierType.java (hunk 4 above)
Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/user/model/IdentifierType.java]: moved to IdentifierUtils

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/filters/ABPMJAYIdFilter.java  4
- src/main/java/in/projecteka/consentmanager/user/IdentifierUtils.java  1
- src/main/java/in/projecteka/consentmanager/user/filters/NameFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/filters/YOBFilter.java  2
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

File: src/main/java/in/projecteka/consentmanager/clients/ClientError.java (hunk 1 above)
Comment (gayatrijagtap) [reply src/main/java/in/projecteka/consentmanager/clients/ClientError.java]: sending NO_PATIENT_FOUND for both zero or multiple patients found cases.

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/filters/NameFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/IdentifierUtils.java  1
- src/main/java/in/projecteka/consentmanager/user/filters/YOBFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/filters/ABPMJAYIdFilter.java  4
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/filters/ABPMJAYIdFilter.java  4
- src/main/java/in/projecteka/consentmanager/user/IdentifierUtils.java  1
- src/main/java/in/projecteka/consentmanager/user/filters/YOBFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/filters/NameFilter.java  2
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

**Comment by ganesanarun** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- src/main/java/in/projecteka/consentmanager/user/filters/ABPMJAYIdFilter.java  4
- src/main/java/in/projecteka/consentmanager/user/filters/NameFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/filters/YOBFilter.java  2
- src/main/java/in/projecteka/consentmanager/user/IdentifierUtils.java  1
         

Clones added
============
- src/test/java/in/projecteka/consentmanager/user/UserServiceTest.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/ProjectEKA/consent-manager/pullRequest?prid=5568924&bid=18262750)

**Pull request merged**